fix: upgrade hcloud-go to v2

jooola opened the pull request:
Attempt to migrate the hcloud-go library to v2, to be able to finally drop the v1 version.

We previously discussed of not going this route, but the effort of porting new features in both hcloud-go versions is bigger than upgrade the library in the terrafrom plugin.

If this change ends up being impossible to test and too dangerous, we will just continue the route we initially choose: migrating for to the plugin framework then the hcloud-go library.

Related to #877

jooola committed:
fix: upgrade hcloud-go to v2

jooola committed:
refactor: fix hcloud-go deprecations

jooola committed:
test: update testing ById func signature

jooola committed:
refactor: update import url to v2

jooola committed:
refactor: update utils to support hcloud-go v2

jooola committed:
refactor: cast and format new int64 IDs (high confidence)

jooola committed:
casting that make sens but might be unneeded

jooola committed:
casts for server resource

jooola committed:
refactor: test and deduplicate ToPublicNetField function

jooola committed:
casts for firewall attachment resource

apricote commented on internal/util/resourceutil/id.go:
```diff
@@ -27,9 +27,9 @@ func ParseID(value types.String) (int, diag.Diagnostics) {
 			"The current version of the provider has problems with IDs > 32bit on 32 bit systems. If possible, switch to a 64 bit system for now. See https://github.com/hetznercloud/hcloud-go/issues/263")
 	}
-	return int(id), diagnostics
+	return id, diagnostics
```
The check above this line (`id > math.MaxInt32 && strconv.IntSize == 32`) can be removed, it was only relevant for the `int(id)` cast.

apricote commented on internal/util/id.go:
```diff
@@ -0,0 +1,25 @@
+	}
+}
+
+func FormatID[T ~int | ~int32 | ~int64](v T) string {
```
Is `~int32` actually used anywhere? This would always result in a runtime panic in `CastInt64()`, so maybe returning a compile-time error instead would be easier to spot?

apricote commented on internal/util/id.go:
```diff
@@ -0,0 +1,25 @@
+}
+
+func ParseID(v string) (int64, error) {
+	return strconv.ParseInt(v, 10, 32)
```
Not sure where this is used yet, but why limit it to `32` bits?

apricote reviewed (commented):
Looking through it I am hopeful for the future of `hcloud-go`!

I am a bit worried about the naked `int()` castings to downcast for terraform on 32 bit platforms. Go just does it, resulting in bad data. What do you think about adding a reverse function that converts from int64 to int on 64-bit platforms, and errors on 32-bit platforms?

jooola committed:
fix: review fixes

jooola committed:
fix: replace naked int casts with wrapper to handle data loss

jooola commented:
> I am a bit worried about the naked int() castings to downcast for terraform on 32 bit platforms. Go just does it, resulting in bad data. What do you think about adding a reverse function that converts from int64 to int on 64-bit platforms, and errors on 32-bit platforms?

Yup, that's a sensible idea. I addressed your review comments and implement this for the relevant casts.

jooola committed:
update hcloud-go to latest 2.19.1

apricote reviewed (approved):
Very nice :rocket: This looks good as far as I can tell, happy to finally put hcloud-go v1 to bed :)
